### PR TITLE
Fix VAC crew post-cutoff pricing lag with rate recalc fallback

### PR DESCRIPTION
**Pull request by JFlo21**
## Summary
Fixes a silent pricing bug where VAC crew and specialized work items (vacuum switches, softswitches, switched banks) retained incorrect SmartSheet prices during post-cutoff rate recalculation. The root cause was that the old CSV's `Compatible Unit Group` column contains verbose group names that don't exist as keys in the new contract rates table, causing the recalc to silently fall through with only a DEBUG log.

## Key Changes

- **SDK workaround**: Added re-export of missing retryable exception classes (`RateLimitExceededError`, `UnexpectedErrorShouldRetryError`, etc.) from `smartsheet.exceptions` into `smartsheet.smartsheet` module to work around a bug in smartsheet-python-sdk 3.8.0 where internal exception lookup fails.

- **Rate recalc fallback logic**: Modified `recalculate_row_price()` to fall back to direct CU-code lookup in the rates table when the mapped group is absent. Only activates on exact match, preventing mis-application of rates.

- **Improved logging visibility**: 
  - Elevated "group not found" logs from DEBUG to WARNING when neither the mapped group nor the CU code exists in new rates
  - Added per-row WARNING logs for dropped VAC crew/helper rows with zero/blank prices
  - Added per-sheet rate recalc summary showing count of recalculated vs. skipped rows and top CUs that couldn't be priced

- **Tracking and diagnostics**: Added counters (`sheet_rate_recalc_counts`, `sheet_rate_recalc_skipped_cus`) to track post-cutoff rate recalc outcomes per sheet, surfacing which CU codes are missing from the new rates CSV.

- **Test coverage**: Added two regression tests:
  - `test_cu_direct_fallback_when_mapped_group_absent_from_new_rates`: Verifies CU-direct fallback works
  - `test_silent_fallthrough_when_neither_group_nor_cu_in_new_rates`: Ensures SmartSheet price is retained when no rate match exists

## Implementation Details

The fix is additive and production-safe:
1. The CU-direct fallback only activates when the mapped group is missing, and only on exact CU match
2. When both group and CU lookups fail, the original SmartSheet price is retained (no invented rates)
3. WARNING logs now surface the specific CU codes and quantities that couldn't be priced, enabling operators to identify missing entries in `NEW_RATES_CSV`
4. Per-sheet summaries make it immediately visible when rate recalc skips occur, rather than silently disappearing into counters

This addresses the known VAC crew pricing-lag signal where post-cutoff rows kept old prices due to missing rate table entries.

https://claude.ai/code/session_01J8i3CGgm2cqd4ReeSP9o8S

---

> [!NOTE]
> **Medium Risk**
> Touches production billing price recalculation and adds a runtime monkeypatch for the Smartsheet SDK, so incorrect rate application or unexpected SDK behavior could impact generated billing outputs; changes are guarded and covered by new tests.
> 
> **Overview**
> Fixes a post-cutoff pricing edge case where rows could silently keep SmartSheet prices when the CU→group mapping produced verbose group names not present in the new contract rate table.
> 
> `recalculate_row_price()` now centralizes CU resolution via `_resolve_cu_code()`, falls back to direct CU-key lookup when the mapped group is missing from `rates_dict`, and reports explicit outcomes (e.g. `missing_rate`, `invalid_quantity`, `zero_rate`) for callers to surface actionable skips.
> 
> Sheet processing adds per-row/per-sheet WARNING visibility for skipped post-cutoff recalcs (including top skipped CUs), and includes a small Smartsheet SDK workaround that re-exports retryable exception classes to restore proper retry behavior. Test coverage is expanded with regressions for the CU-direct fallback, safety retention when no match exists, and `_resolve_cu_code` behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 477938dfe389f9b9a7301c978c2994ee6836523b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by JFlo21** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/jflo21/pr-checks/70ffb2ce-26ca-4cca-a708-a0e6300e5be4?refs=477938dfe389f9b9a7301c978c2994ee6836523b&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/jflo21/project/eb083177-20b4-46e7-9447-b20c1e497e41/pr-check/c2b9581f-9ea6-4797-aeaf-2663d36f6cb9?refs=477938dfe389f9b9a7301c978c2994ee6836523b&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: CLAUDE.md [hunk 1]
```diff
@@ -349,3 +349,50 @@ When proposing new workflows, dynamically evaluate the absolute best technology.
...
+  fall-through* in `recalculate_row_price()`, not a cutoff-column
+  bug. **Context:** The cutoff rule is correctly keyed on
+  `Snapshot Date >= RATE_CUTOFF_DATE` at
+  `generate_weekly_pdfs.py:2127`; Weekly Reference Logged Date is
+  the wrong column for this check and must NOT be substituted —
+  operators depend on Snapshot Date semantics. **Real root cause:**
+  `build_cu_to_group_mapping()` reads the old CSV's
+  `Compatible Unit Group` column, which mixes short codes (e.g.
+  `ANC-M`, `CPD-SW`) with verbose names (e.g. `Vacuum Switch`,
+  `Overhead Switching"`, `Softswitch Type K"`, `1200 KVAR Switched
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply CLAUDE.md:362]: Those trailing quotes are faithful to the raw values in `CU List - Corpus North & South.csv` — they're CSV escaping artifacts that live in the actual "Compatible Unit Group" column (e.g. `Overhead Switching"`, `Softswitch Type K"`, `Switch Bank Assby"`). Stripping them from the ledger would make it harder for the next engineer to `grep` the CSV for those literal strings while debugging a "missing rate" WARNING. Leaving as-is so the ledger matches the data on disk.

---
_Generated by [Claude Code](https://claude.ai/code)_
